### PR TITLE
[FIX] use DP attention group for MoonViT reductions

### LINKED ISSUE #34442: [Bug] Kimi-VL TP4/DP2 + DP-attention deadlocks on first image request (MoonViT encoder reduces on the global TP group)
### Checklist
- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version. (Verified on the 2026-08-10 nightly snapshot of main, `7c90840b`; the implicated code is unchanged on main as of filing.)
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

Serving Kimi-VL with `--enable-dp-attention`, more than one DP group, and `attn_tp_size > 1` (i.e. `tp_size > dp_size`; tested at TP=4/DP=2): the **first request containing an image wedges all scheduler ranks**. Every subsequent request stalls; after ~430 s all ranks log `Scheduler watchdog timeout (watchdog_timeout=300, soft=False)` and the engine is torn down via SIGQUIT. No exception is raised; it is a distributed deadlock, not a crash. TP=2/DP=2 layouts have `attn_tp_size == 1`, so MoonViT is unsharded and performs no reduction; they should be unaffected.

Reproduced 10/10 across two independent 4×H100 machines, in both control-broadcast modes, and regardless of which DP group round-robin assigns the image to. With default warmup the server never becomes healthy at all: the built-in startup warmup request includes an image for multimodal models, so the same wedge fires before the server is marked ready (`/health` 503 until teardown). With `--skip-server-warmup`, `/health` is 200 and text requests complete in ~0.3 s until the first image request, after which health checks time out.

| container image / version | config | result |
|---|---|---|
| nightly `7c90840b` (2026-08-10) | `--tp 4 --dp 2 --enable-dp-attention` | **deadlock on 1st image request** (text OK) |
| nightly `7c90840b` | same + `--enable-dp-attention-local-control-broadcast` | **deadlock on 1st image request** |
| nightly `7c90840b` | `--tp 4` (no DP-attention) | clean (240 s image load test, 0 errors) |
| v0.5.13 | `--tp 4 --dp 2 --enable-dp-attention` (both ctrl modes) | clean (13,432 mixed requests over 2×480 s on box 1; 23 image prefills spread across both DP groups on box 2) |

### Likely root cause (source-verified at `7c90840b`)

PR #30869 ("fix: fix Kimi-VL encoder parallelism", merged 2026-07-14) made MoonViT tensor-parallel. The sharding and the reduction disagree about the communicator:

- `kimi_vl_moonvit.py:344` shards MoonViT's `ColumnParallelLinear`/`RowParallelLinear` with `tp_size = get_parallel().attn_tp_size` (TP is the default path when `--mm-enable-dp-encoder` is not set).
- But the reduction group is selected by the `use_dp_attention_reduce` constructor flag (`linear.py:1432`): with it, `attn_tp_group.all_reduce`; without it, `tensor_model_parallel_all_reduce` on the **full global TP group**. MoonViT omits the flag at two sites: the MLP `fc1` `RowParallelLinear` and `VisionAttention`'s output projection. Qwen3-VL (`qwen3_vl.py:137,211,300`) and Kimi-K2.5 (`kimi_k25.py:130`) pass `use_dp_attention_reduce=is_dp_attention_enabled()` at the equivalent sites, which makes the likely fix immediately obvious.
- Under DP-attention only the image-owning group executes MoonViT, so the two DP groups no longer enqueue the same ordered sequence on the global TP communicator; the sequences are permanently offset from the first vision-side reduce onward, and the communicator wedges.
- v0.5.13's MoonViT used plain replicated `nn.Linear` (no vision-side collectives), which is why v0.5.13 serves the identical workload cleanly.

A/B test: adding `--mm-enable-dp-encoder` (switches MoonViT to its `use_data_parallel` path, removing the TP reductions, all else equal) **eliminates the deadlock**: same image, same box, same nightly; 5 consecutive image requests served, `/health` 200 afterwards. This strongly supports the mechanism and provides an immediate workaround. Any other VLM whose encoder TP-shards with `attn_tp` sizing but reduces via the default global-group path may be affected; I only tested Kimi-VL.

### Reproduction

```
python3 -m sglang.launch_server \
  --model-path moonshotai/Kimi-VL-A3B-Instruct \
  --tp 4 --dp 2 --enable-dp-attention \
  --skip-server-warmup \
  --trust-remote-code --host 0.0.0.0 --port 8000
```

```python
import base64, io, json, urllib.request
from PIL import Image
import numpy as np

def post(body):
    req = urllib.request.Request("http://127.0.0.1:8000/v1/chat/completions",
        data=json.dumps(body).encode(), headers={"Content-Type": "application/json"})
    with urllib.request.urlopen(req, timeout=120) as r:
        return r.read(200)

print(post({"model": "default", "max_tokens": 16,
            "messages": [{"role": "user", "content": "Count to three."}]}))  # succeeds

rng = np.random.default_rng(0)
arr = (rng.random((800, 800, 3)) * 255).astype("uint8")
buf = io.BytesIO(); Image.fromarray(arr).save(buf, format="PNG")
url = "data:image/png;base64," + base64.b64encode(buf.getvalue()).decode()
print(post({"model": "default", "max_tokens": 24,
            "messages": [{"role": "user", "content": [
                {"type": "image_url", "image_url": {"url": url}},
                {"type": "text", "text": "Describe this image."}]}]}))  # times out after 120 s client-side; engine remains wedged
```

### Per-rank stacks at the wedge

py-spy cannot attach in these containers (ptrace restricted), so stacks were captured via `PYTHONFAULTHANDLER=1` + `SIGABRT` on one scheduler per repro run. Ranks of the idle DP group are back in the receive loop, blocked in the control broadcast on the global TP CPU group, waiting for ranks that never return:

```
File "torch/distributed/distributed_c10d.py", line 3151 in broadcast
File ".../srt/utils/common.py", line 2256 in broadcast_pyobj
File ".../srt/managers/scheduler_components/request_receiver.py", line 202 in _broadcast_reqs_across_ranks
File ".../srt/managers/scheduler_components/request_receiver.py", line 93 in recv_requests
File ".../srt/managers/scheduler.py", line 1738 in event_loop_overlap
```

Ranks of the image-owning group are inside the model forward, past the vision encoder (via `mm_utils.py:730 general_mm_embed_routine → kimi_vl.py:207`), blocked at a downstream kernel launch (`dp_attention.py:826 dp_scatter` on one machine, a GEMM on the other). Those frames are consequences, not the blocked collective: NCCL enqueues are asynchronous, so the CPU proceeds past MoonViT's unpaired all-reduce and blocks at whatever launch it reaches once the stream stalls.

### Environment

<details><summary>python3 -m sglang.check_env</summary>

```
Python: 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0]
CUDA available: True
GPU 0,1,2,3: NVIDIA H100 PCIe
GPU 0,1,2,3 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 580.159.04
PyTorch: 2.13.0+cu130
sglang: 0.0.0.dev1+g7c90840ba
sglang-kernel: 0.4.6.post1
flashinfer_python: 0.6.15.post1
flashinfer_jit_cache: 0.6.15.post1+cu130
triton: 3.7.1
transformers: 5.12.1
torchao: 0.17.0+cu130
numpy: 2.3.5
pyzmq: 27.1.0
xgrammar: 0.2.1
torchcodec: 0.15.0+cu130
NVIDIA Topology: 
	GPU0	GPU1	GPU2	GPU3	NIC0	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	NV12	SYS	SYS	NODE	0-47,96-143	0		N/A
GPU1	NV12	 X 	SYS	SYS	NODE	0-47,96-143	0		N/A
GPU2	SYS	SYS	 X 	NODE	SYS	48-95,144-191	1		N/A
GPU3	SYS	SYS	NODE	 X 	SYS	48-95,144-191	1		N/A
ulimit soft: 1024
```

</details>

Model: `moonshotai/Kimi-VL-A3B-Instruct` (bf16). Reproduced on a second independent 4×H100 machine (driver 580.95.05). Working comparison: `lmsysorg/sglang:v0.5.13`, identical hardware and flags.

### Related

- **Why existing coverage misses this:** the registered Kimi-VL DP-attention image test (`test/registered/dp_attn/test_dp_attention.py`) runs `tp2 + dp2`, i.e. `attn_tp_size == 1`, where MoonViT is unsharded and performs no reductions. The manual TP=4/DP=2 VLM test (`test_dp_attention_large.py::TestDPAttentionDP2TP4VLM`) uses Qwen3-VL, whose encoder routes reductions correctly. No test runs Kimi-VL with `attn_tp_size > 1` and an image.
- Found while retesting #29010 (unrelated shm race; its fix #29543 is not implicated, since the wedge occurs downstream of shm unwrapping).

### PR DESCRIPTION
**Pull request by Jad-yehya**
## Motivation

Kimi-VL can deadlock on the first image request when DP attention is enabled with more than one DP group and `attn_tp_size > 1`, such as TP4/DP2.

MoonViT shards its attention output projection and MLP output projection across the attention TP group, but these layers reduce through the global TP group. Only the DP group handling the image executes these collectives, so ranks enqueue different collective sequences.

Fixes #34442.

## Modifications

- Use the DP-attention TP group for the MoonViT attention output projection.
- Use the DP-attention TP group for the MoonViT MLP output projection.
- Add a unit test covering both reduction paths.

## Accuracy Tests

The change only selects the communicator for existing reductions.
The added unit test checks both projections.

## Speed Tests and Profiling

Not run. The change does not add collectives or alter tensor shapes; it routes two existing reductions to the attention TP group.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #32713693655](https://github.com/sgl-project/sglang/actions/runs/32713693655)
Latest PR Test (Extra): :x: [Run #32713693404](https://github.com/sgl-project/sglang/actions/runs/32713693404)
Latest PR Test (AMD ROCm 7.2): :x: [Run #32713693752](https://github.com/sgl-project/sglang/actions/runs/32713693752)

### CODE REVIEW

**Comment by Jad-yehya** [comment]
@mickqian  @yuan-luo  @yhyang201  could one of you review?
